cmd/juju/application/resolved_test.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2012, 2013 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package application_test
+
+import (
+	"github.com/juju/cmd/cmdtesting"
+	"github.com/juju/testing"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/cmd/juju/application"
+)
+
+type ResolvedSuite struct {
+	testing.IsolationSuite
+
+	mockAPI *mockResolveAPI
+}
+
+func (s *ResolvedSuite) SetUpTest(c *gc.C) {
+	s.IsolationSuite.SetUpTest(c)
+	s.mockAPI = &mockResolveAPI{Stub: &testing.Stub{}}
+}
+
+var _ = gc.Suite(&ResolvedSuite{})
+
+func (s *ResolvedSuite) runResolved(c *gc.C, args []string) error {
+	cmd := application.NewResolvedCommandForTest(s.mockAPI, s.mockAPI)
+	_, err := cmdtesting.RunCommand(c, cmd, args...)
+	return err
+}
+
+var resolvedTests = []struct {
+	args       []string
+	err        string
+	apiVersion int
+	retry      bool
+	all        bool
+	legacyUnit string
+	units      []string
+}{
+	{
+		err: `no unit specified`,
+	}, {
+		args: []string{"jeremy-fisher"},
+		err:  `unit name "jeremy-fisher" not valid`,
+	}, {
+		args: []string{"jeremy-fisher/99", "--all"},
+		err:  `specifying unit names\(s\) with --all not supported`,
+	}, {
+		apiVersion: 5,
+		args:       []string{"--all"},
+		err:        `resolving all units or more than one unit not support by this version of Juju`,
+	}, {
+		apiVersion: 5,
+		args:       []string{"jeremy-fisher/98", "jeremy-fisher/99"},
+		err:        `resolving all units or more than one unit not support by this version of Juju`,
+	}, {
+		apiVersion: 5,
+		args:       []string{"jeremy-fisher/99"},
+		legacyUnit: "jeremy-fisher/99",
+		retry:      true,
+	}, {
+		apiVersion: 5,
+		args:       []string{"jeremy-fisher/99", "--no-retry"},
+		legacyUnit: "jeremy-fisher/99",
+	}, {
+		args:  []string{"jeremy-fisher/98", "jeremy-fisher/99", "--no-retry"},
+		units: []string{"jeremy-fisher/98", "jeremy-fisher/99"},
+	}, {
+		args:  []string{"jeremy-fisher/98", "jeremy-fisher/99"},
+		units: []string{"jeremy-fisher/98", "jeremy-fisher/99"},
+		retry: true,
+	}, {
+		args: []string{"--all", "--no-retry"},
+		all:  true,
+	},
+}
+
+func (s *ResolvedSuite) TestResolved(c *gc.C) {
+	for i, t := range resolvedTests {
+		s.mockAPI.ResetCalls()
+		c.Logf("test %d: %v", i, t.args)
+		if t.apiVersion > 0 {
+			s.mockAPI.version = t.apiVersion
+		} else {
+			s.mockAPI.version = 6
+		}
+		err := s.runResolved(c, t.args)
+		if t.err != "" {
+			c.Assert(err, gc.ErrorMatches, t.err)
+			continue
+		} else {
+			c.Assert(err, jc.ErrorIsNil)
+		}
+		if t.legacyUnit != "" && !t.all {
+			s.mockAPI.CheckCallNames(c, "BestAPIVersion", "Resolved", "Close", "Close")
```

Comment:
Why two closes?

Reply:
One for Client facade and one for Application facade

Reply:
Seems a bit strange, and is a side effect of our mocks. Should be documented.

Reply:
We open both facades (in this case). When each facade is opened, we do a defer Close(). The mocks are accurately recording the expected api calls here.